File: bindings/go/backend/lease.go
package backend

import (
	"context"
	"time"
)

type LeaseDefinitionID struct {
	Name    string
	Version uint64
}

type LeaseDefinition struct {
	ID              LeaseDefinitionID
	ResourcePattern string
	TTL             time.Duration
	RenewEvery      time.Duration
	MaxTTL          *time.Duration
	FencingToken    bool
}

type LeaseRegisterDefinitionResult struct {
	RegisteredNew bool
	Definition    LeaseDefinition
}

type LeaseRecord struct {
	DefinitionID LeaseDefinitionID
	Resource     string
	Holder       *string
	ExpiresAt    time.Time
	FencingToken uint64
}

type LeaseAcquireRequest struct {
	DefinitionID LeaseDefinitionID
	Resource     string
	Holder       string
	Now          time.Time
}

type LeaseRenewRequest struct {
	DefinitionID LeaseDefinitionID
	Resource     string
	Holder       string
	FencingToken uint64
	Now          time.Time
}

type LeaseReleaseRequest struct {
	DefinitionID LeaseDefinitionID
	Resource     string
	Holder       string
	FencingToken uint64
}

type LeaseInspectRequest struct {
	DefinitionID LeaseDefinitionID
	Resource     string
}

type LeaseAcquireResult struct {
	Acquired bool
	Lease    *LeaseRecord
}

type LeaseStore interface {
	RegisterDefinition(ctx context.Context, backend Backend, definition LeaseDefinition) (LeaseRegisterDefinitionResult, error)

	RegisterDefinitionTx(ctx context.Context, tx Transaction, definition LeaseDefinition) (LeaseRegisterDefinitionResult, error)

	InspectDefinition(ctx context.Context, backend Backend, definitionID LeaseDefinitionID) (*LeaseDefinition, error)

	InspectDefinitionTx(ctx context.Context, tx Transaction, definitionID LeaseDefinitionID) (*LeaseDefinition, error)

	Acquire(ctx context.Context, backend Backend, request LeaseAcquireRequest) (LeaseAcquireResult, error)

	AcquireTx(ctx context.Context, tx Transaction, request LeaseAcquireRequest) (LeaseAcquireResult, error)

	Renew(ctx context.Context, backend Backend, request LeaseRenewRequest) (LeaseRecord, error)

	RenewTx(ctx context.Context, tx Transaction, request LeaseRenewRequest) (LeaseRecord, error)

	Release(ctx context.Context, backend Backend, request LeaseReleaseRequest) error

	ReleaseTx(ctx context.Context, tx Transaction, request LeaseReleaseRequest) error

	Inspect(ctx context.Context, backend Backend, request LeaseInspectRequest) (*LeaseRecord, error)

	InspectTx(ctx context.Context, tx Transaction, request LeaseInspectRequest) (*LeaseRecord, error)
}

File: bindings/java/com/statespec/backend/Lease.java
package com.statespec.backend;

import com.statespec.backend.Backend.BackendException;
import com.statespec.backend.Backend.Transaction;

import java.time.Duration;
import java.time.Instant;
import java.util.Optional;

public interface Lease {
    record LeaseDefinitionId(
        String name,
        long version
    ) {}

    record LeaseDefinition(
        LeaseDefinitionId id,
        String resourcePattern,
        Duration ttl,
        Duration renewEvery,
        Optional<Duration> maxTtl,
        boolean fencingToken
    ) {}

    record LeaseRegisterDefinitionResult(
        boolean registeredNew,
        LeaseDefinition definition
    ) {}

    record LeaseRecord(
        LeaseDefinitionId definitionId,
        String resource,
        Optional<String> holder,
        Instant expiresAt,
        long fencingToken
    ) {}

    record LeaseAcquireRequest(
        LeaseDefinitionId definitionId,
        String resource,
        String holder,
        Instant now
    ) {}

    record LeaseRenewRequest(
        LeaseDefinitionId definitionId,
        String resource,
        String holder,
        long fencingToken,
        Instant now
    ) {}

    record LeaseReleaseRequest(
        LeaseDefinitionId definitionId,
        String resource,
        String holder,
        long fencingToken
    ) {}

    record LeaseInspectRequest(
        LeaseDefinitionId definitionId,
        String resource
    ) {}

    record LeaseAcquireResult(
        boolean acquired,
        Optional<LeaseRecord> lease
    ) {}

    LeaseRegisterDefinitionResult registerDefinition(
        Backend backend,
        LeaseDefinition definition
    ) throws BackendException;

    LeaseRegisterDefinitionResult registerDefinitionTx(
        Transaction tx,
        LeaseDefinition definition
    ) throws BackendException;

    Optional<LeaseDefinition> inspectDefinition(
        Backend backend,
        LeaseDefinitionId definitionId
    ) throws BackendException;

    Optional<LeaseDefinition> inspectDefinitionTx(
        Transaction tx,
        LeaseDefinitionId definitionId
    ) throws BackendException;

    LeaseAcquireResult acquire(
        Backend backend,
        LeaseAcquireRequest request
    ) throws BackendException;

    LeaseAcquireResult acquireTx(
        Transaction tx,
        LeaseAcquireRequest request
    ) throws BackendException;

    LeaseRecord renew(
        Backend backend,
        LeaseRenewRequest request
    ) throws BackendException;

    LeaseRecord renewTx(
        Transaction tx,
        LeaseRenewRequest request
    ) throws BackendException;

    void release(
        Backend backend,
        LeaseReleaseRequest request
    ) throws BackendException;

    void releaseTx(
        Transaction tx,
        LeaseReleaseRequest request
    ) throws BackendException;

    Optional<LeaseRecord> inspect(
        Backend backend,
        LeaseInspectRequest request
    ) throws BackendException;

    Optional<LeaseRecord> inspectTx(
        Transaction tx,
        LeaseInspectRequest request
    ) throws BackendException;
}
